Sources/EdgeNetworkHandlers/EdgeHitProcessor.swift

```diff
@@ -0,0 +1,74 @@
+// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+// OF ANY KIND, either express or implied. See the License for the specific language
+// governing permissions and limitations under the License.
+//
+
+import AEPServices
+import Foundation
+
+/// A `HitProcessing` which handles the processing of `EdgeHit`s
+class EdgeHitProcessor: HitProcessing {
+    private let LOG_TAG = "EdgeHitProcessor"
+    private var networkService: EdgeNetworkService
+    private var networkResponseHandler: NetworkResponseHandler
+    var retryInterval = TimeInterval(5)
+
+    init(networkService: EdgeNetworkService, networkResponseHandler: NetworkResponseHandler) {
+        self.networkService = networkService
+        self.networkResponseHandler = networkResponseHandler
+    }
+
+    func processHit(entity: DataEntity, completion: @escaping (Bool) -> Void) {
+        guard let data = entity.data, let edgeHit = try? JSONDecoder().decode(EdgeHit.self, from: data) else {
+            // can't convert data to hit, unrecoverable error, move to next hit
+            Log.debug(label: LOG_TAG, "processHit - Failed to decode edge hit '\(entity.uniqueIdentifier)'.")
+            completion(true)
+            return
+        }
+
+        let callback: ResponseCallback = NetworkResponseCallback(requestId: entity.uniqueIdentifier, responseHandler: networkResponseHandler)
+        let hitCallback = EdgeHitResponseCallback(completion: completion, callback: callback)
+        networkService.doRequest(url: edgeHit.url,
+                                 requestBody: edgeHit.request,
+                                 requestHeaders: edgeHit.headers,
+                                 responseCallback: hitCallback,
+                                 retryTimes: Constants.Defaults.NETWORK_REQUEST_MAX_RETRIES)
+    }
+
+}
+
+/// A wrapper struct to handle the network service callback and pass it to the customer facing callback
+private struct EdgeHitResponseCallback: ResponseCallback {
+    private let LOG_TAG = "EdgeHitResponseCallback"
+    let completion: (Bool) -> Void
+    let callback: ResponseCallback
+
+    func onResponse(jsonResponse: String) {
+        callback.onResponse(jsonResponse: jsonResponse)
+        completion(true)
+    }
+
+    func onError(jsonError: String) {
+        guard let data = jsonError.data(using: .utf8) else { return }
+        guard let edgeErrorResponse = try? JSONDecoder().decode(EdgeResponse.self, from: data) else {
+            Log.warning(label: LOG_TAG, "onError - The conversion to JSON failed for server error response: \(jsonError).")
+            return
+        }
+
+        callback.onError(jsonError: jsonError)
+        let isRecoverable = edgeErrorResponse.errors?.contains(where: {$0.isRecoverable}) ?? false
+        completion(!isRecoverable) // error, retry this hit if it is recoverable
+    }
+
+    func onComplete() {
+        callback.onComplete()
```

Comment:
I was testing with this code and at some point I was getting an empty response back from Konductor. What I saw was only `onComplete` was called, but not `onResponse` or `onError` which meant the hits in the queue were not being removed. I would suggest adding `completion(true)` here as a fail-safe.

Reply:
Is there any concern with invoking the callback more than once in the case that we do
onResponse() -> onComplete() that would invoke the callback twice, same with onError() -> onComplete().

As a side note, if we moved away from the protocol pattern this would simplify this logic and might make it a bit safer.
https://github.com/adobe/aepsdk-edge-ios/issues/90#issuecomment-715585562

Reply:
Yes, you are right. It's my understanding that `completion()` will remove the top hit in the queue so it would be bad to call it multiple times for the same event. However, `onResponse` and `onError` may not get called if there are errors sending the request but they may get called multiple times when response streaming is enabled and when batching experience events. I believe `onComplete` is always called once, so we should call `completion(true)` only there. Also, `onComplete` should get called once after any retries are attempted, so you don't need to add any retry logic, just pass `true` to `completion`.